### ISSUE TITLE
RFC (all discussion of the RFC should go here)

### ISSUE THREAD

**Issue by phitoduck** [phitoduck/project-generator-cli-poc#1]
Let's discuss the scope and functionality of the RFC in this thread.

We can submit lots of PRs to the `README.md` as ideas stick.

**Comment by phitoduck** [comment]
This is probably a surprising statement, but for my personal preferences, I want the tool to be able to support subfolders with components in different languages.

(Unfortunately, IMO) the Infrastructure as Code world currently is strongest in TypeScript. Pulumi, AWS CDK, CDK TF, and CDK8s have their largest feature sets available in TypeScript, and smaller feature sets in Python, Java, etc.

For most of my projects these days, I end up wanting something like this:

```text
my-project/
    python-package/ # a python project
        src/
        setup.py
        setup.cfg
        pyproject.toml

    iac/ # could be AWS CDK, Terraform, CloudFormation, Pulumi, or a mix... but is often TypeScript
```

A common setup is:

```text
my-streamlit-project/
    .github/
        deploy.yml  # deploy using IaC
        publish-docs.yml
        build.yml # build streamlit app and fastAPI backend for deployment
    streamlit-app/
        src/
        docs/    # in my project, I've been putting both python projects into the same sphinx docs
        setup.py
        setup.cfg
        Dockerfile
    fast-api-server/
        src/
        docs/
        setup.py
        setup.cfg
        Dockerfile
     infrastructure/
         src/
         package.json
         app.ts
     docker-compose.yml
     README.md
     Makefile (or Justfile)
```

**Comment by phitoduck** [comment]
Consideration for CI (especially with monorepos)

Changes to certain areas of code have no effect on other areas. A naive approach to CI is to run all tests, builds, code-quality checks for *every* PR. This has two disadvantages:

1. It burns build minutes, API quota, etc. for changes to small files like the `README.md`
2. Besides cost, it's annoying for developers since longer builds mean slower iteration cycles.

Areas we can optimize our builds:

1. Only run tests if the code that affects those tests have changed.
2. Only build docs if the inputs to the docs have changed.
3. Cache dependencies. No need to re-run `pip install ...` or `yarn install` if the `Pipfile.lock` and `package.lock.json` haven't changed (assuming we have cached dependencies available locally)
4. Run build steps in parallel wherever possible.

Here is an [example of how a rootski contributor added caching to rootski](https://github.com/rootski-io/rootski/blob/trunk/.github/workflows/rootski-ci.yml#L41-L45).

**Comment by Ealameda31** [comment]
> This is probably a surprising statement, but for my personal preferences, I want the tool to be able to support subfolders with components in different languages.

Yes, this caught me off guard. I do not know how far we will go in terms of IAC support, but we can definitely build a python-based tool that checks the components you laid out in the `README.md`.

That has been my biggest gripe with tests for a long time. Many people just run them mindlessly again without conditional logic on whether or not the test **should** be run. This is not too difficult to implement though. We could have an overall script-to-test mapper that Actions checks with every commit (though that might be quite hacky, and we can think of better ways).

People tend to version all their dependencies (i.e.,`numpy==1.22.3`), which can set up their project for deprecation in the future depending on Python and/or library versions. I typically enforce versions on libraries that need them and update my code to keep up-to-date with common libraries  (i.e., `numpy`, `pandas`, `sklearn`, etc.). Caching probably does not work too well here because the library version you cached might not be the latest for the package. Of course we are talking long-term deprecation, but it is something to keep in mind.

**Comment by Ealameda31** [comment]
FWIW, I will probably follow your design philosophy - let's try to keep it consistent and develop features that make sense for it 🧶

**Comment by phitoduck** [comment]
@Ealameda31, you're here! 

I agree with your point about saving IaC for a later as long as we don't make design decisions that would preclude it.

**Comment by phitoduck** [comment]
Capturing benefits of TS for IaC for historical purposes:

The most cutting-edge tools are in TS these days. I keep crossing my fingers that Python will catch up, but that looks a hopelessly long way off 😐. Here are some examples:

- Pulumi addresses many weaknesses of AWS CDK, but doesn't get the support of the AWS Team... so they simply made it so [Pulumi can import CDK constructs (but only in TypeScript)](https://www.pulumi.com/blog/aws-cdk-on-pulumi/).
- [Serverless Stack (sst)](https://sst.dev/) is a framework built on top of AWS CDK that lets you easily compose AWS Lambda functions, API Gateways, Queues, etc. But the most amazing feature is that it let's you deploy those things to AWS for development *instantly*--as in, you hit `CTRL` + `S`, and in under a second your changes are deployed. This is a game changer for the serverless development experience.
- Certain advanced features for AWS CDK are only available in TypeScript
    - writing a construct once (in TypeScript), then *generating* libraries for that construct in Python, Java, Go, etc. with almost no extra effort
    - Advanced "CDK Escape Hatches" such as ["mocking" intrinsic functions like `Fn::Import`](https://github.com/phitoduck/private-construct-hub/blob/trunk/src/main.ts#L36-L51) on L1 constructs.
- Many community/official AWS CDK constructs are only available in TypeScript 😐 (private instance of Construct Hub being on of them)
- Even [`projen`](https://github.com/projen/projen) (a project template generator library by AWS) is made for TypeScript first, with the ability to generate bindings in other languages such as Python.

You can absolutely still write IaC in Python, but at least for my team at work, it's felt like we've been swimming against the flow of the IaC community as we've resisted adopting TS.

This is all to say, I think polyglot repositories should be possible--certainly not mandatory. I expect many users would choose to go with Python for everything and that's fine.

**Comment by Ealameda31** [comment]
You are going to make me a fan of TypeScript in this journey, aren't you? I will give it a look but Terraform + Python works well for me usually. 

Of course I am here! 💯 I will try to submit a PR for this sometime this weekend to get us going (probably something simple). 

In the meantime, I have a web-application at work that is giving me trouble - somehow dependencies went awry 🤷 this is why we need to migrate it to K8s...

**Comment by Ealameda31** [comment]
Is it just us two to start?

**Comment by phitoduck** [comment]
Sorry, I totally thought I answered this. So far yes, it's just the two of us. Although, my team at work is going to need to solve this problem within the next month or two. I'm hoping that what we build here will be extensible enough that I can simply write a plugin to it for my team at work. We will likely get some contributions from my team at that point if they agree to use the tool.

We could certainly recruit more as well. Once we establish the basis of our framework, it should be easier for others to see how it is meant to be extended, so we may want to wait until get the basics lined up.

**Comment by phitoduck** [comment]
I created a PR with a POC in an attempt to communicate what I've learned about how `projen` approaches this same problem. #2.
